plonky2/src/recursion/cyclic_recursion.rs

```diff
@@ -0,0 +1,490 @@
+    fn from_slice<F: RichField + Extendable<D>, C: GenericConfig<D, F = F>, const D: usize>(
+        slice: &[Target],
+        common_data: &CommonCircuitData<F, C, D>,
+    ) -> Result<Self> {
+        let cap_len = common_data.config.fri_config.num_cap_elements();
+        let len = slice.len();
+        ensure!(len >= 4 + 4 * cap_len + 1, "Not enough public inputs");
+        let base_case = BoolTarget::new_unsafe(slice[len - 1]);
+        let constants_sigmas_cap = MerkleCapTarget(
+            (0..cap_len)
+                .map(|i| HashOutTarget {
+                    elements: std::array::from_fn(|j| slice[len - 1 - 4 * (cap_len - i) + j]),
+                })
+                .collect(),
+        );
+        let circuit_digest = HashOutTarget {
+            elements: std::array::from_fn(|i| slice[len - 5 - 4 * cap_len + i]),
+        };
+
+        Ok(Self {
+            circuit_digest,
+            constants_sigmas_cap,
+            base_case,
+        })
+    }
+}
+
+impl<F: RichField + Extendable<D>, const D: usize> CircuitBuilder<F, D> {
+    pub fn cyclic_recursion<C: GenericConfig<D, F = F>>(
+        mut self,
+        previous_virtual_public_inputs: &[Target],
+        previous_base_case: Target,
+        mut common_data: CommonCircuitData<F, C, D>,
+    ) -> Result<(CircuitData<F, C, D>, CyclicRecursionTarget<D>)>
+    where
+        C::Hasher: AlgebraicHasher<F>,
+        [(); C::Hasher::HASH_SIZE]:,
+    {
+        ensure!(
+            previous_virtual_public_inputs.len() == self.num_public_inputs(),
+            "Incorrect number of public inputs."
+        );
+
+        let verifier_data = VerifierCircuitTarget {
+            constants_sigmas_cap: self.add_virtual_cap(self.config.fri_config.cap_height),
+            circuit_digest: self.add_virtual_hash(),
+        };
+        // The verifier data are public inputs.
+        self.register_public_inputs(&verifier_data.circuit_digest.elements);
+        for i in 0..self.config.fri_config.num_cap_elements() {
+            self.register_public_inputs(&verifier_data.constants_sigmas_cap.0[i].elements);
+        }
+
+        let dummy_verifier_data = VerifierCircuitTarget {
+            constants_sigmas_cap: self.add_virtual_cap(self.config.fri_config.cap_height),
+            circuit_digest: self.add_virtual_hash(),
+        };
+
+        // Flag set to true for the base case of the cycle where we verify a dummy proof to bootstrap the cycle. Set to false otherwise.
+        // Unsafe is ok since `base_case` is a public input and its booleaness should be checked in the verifier.
+        let base_case = self.add_virtual_bool_target_unsafe();
+        self.register_public_input(base_case.target);
+
+        common_data.num_public_inputs = self.num_public_inputs();
+        // The `conditionally_verify_proof` gadget below takes 2^12 gates, so `degree_bits` cannot be smaller than 13.
+        common_data.fri_params.degree_bits = common_data.fri_params.degree_bits.max(13);
```

Comment:
Hm it seems a little odd to me that the user has to guess these params, but we automatically correct certain mistakes. Might be best to throw an error pointing out the mistakes, to avoid any confusion. Like a user might think they got it to work with 2^12, when really it was bumped to 2^13 (if they don't check the logs for the real size).

Reply:
Good point, I just remove it altogether since it's caught in the `CommonCircuitData` equality check.